File: src/infra/tmux.ts
import { runProcess } from './process'

export type TmuxClient = {
  run: (args: string[]) => Promise<void>
  runQuiet: (args: string[]) => Promise<void>
  capture: (args: string[]) => Promise<string>
}

export type PaneGeometry = {
  width: number
  height: number
}

export type PaneContext = PaneGeometry & {
  paneId: string
  currentPath: string
}

export type PaneStartContext = PaneContext & {
  inCopyMode: boolean
}

export type WindowPaneContext = PaneStartContext & {
  left: number
  top: number
  active: boolean
}

export type PaneBorderLines =
  | 'single'
  | 'double'
  | 'heavy'
  | 'simple'
  | 'number'
  | 'spaces'

export type DisplayPopupOptions = PaneGeometry & {
  command: string[]
  currentPath: string
  targetClient: string
  targetPane: string
  x?: number | string
  y?: number | string
}

export const createTmuxClient = (): TmuxClient => ({
  async run(args) {
    await runProcess('tmux', args)
  },
  async runQuiet(args) {
    try {
      await runProcess('tmux', args)
    } catch {
      // Ignore cleanup failures.
    }
  },
  async capture(args) {
    const result = await runProcess('tmux', args)
    return result.stdout
  },
})

export const focusClientPane = async (
  tmux: TmuxClient,
  paneId: string,
  clientTty: string,
): Promise<void> => {
  try {
    await tmux.run(['switch-client', '-c', clientTty, '-t', paneId])
  } catch (error) {
    throw new Error('tmux-fuzzy-motion: client not found', { cause: error })
  }
}

export const enterCopyMode = async (
  tmux: TmuxClient,
  paneId: string,
): Promise<void> => {
  try {
    await tmux.run(['copy-mode', '-t', paneId])
  } catch (error) {
    throw new Error('tmux-fuzzy-motion: failed to enter copy-mode', {
      cause: error,
    })
  }
}

export const getPaneStartContext = async (
  tmux: TmuxClient,
  paneId: string,
): Promise<PaneStartContext> => {
  let output = ''

  try {
    output = (
      await tmux.capture([
        'display-message',
        '-p',
        '-t',
        paneId,
        '#{pane_id}\t#{pane_in_mode}\t#{pane_width}\t#{pane_height}\t#{pane_current_path}',
      ])
    ).trim()
  } catch (error) {
    throw new Error('tmux-fuzzy-motion: pane not found', { cause: error })
  }

  const [resolvedPaneId, paneInMode, width, height, currentPath] =
    output.split('\t')

  if (
    !resolvedPaneId ||
    !paneInMode ||
    !currentPath ||
    [width, height]
      .map((value) => Number(value))
      .some((value) => !Number.isFinite(value))
  ) {
    throw new Error('tmux-fuzzy-motion: failed to resolve pane context')
  }

  return {
    paneId: resolvedPaneId,
    inCopyMode: paneInMode === '1',
    width: Number(width),
    height: Number(height),
    currentPath,
  }
}

export const displayPopup = async (
  tmux: TmuxClient,
  options: DisplayPopupOptions,
): Promise<void> => {
  await tmux.run([
    'display-popup',
    '-E',
    '-B',
    '-c',
    options.targetClient,
    '-t',
    options.targetPane,
    '-d',
    options.currentPath,
    '-x',
    String(options.x ?? '#{popup_pane_left}'),
    '-y',
    String(options.y ?? '#{popup_pane_top}'),
    '-w',
    String(options.width),
    '-h',
    String(options.height),
    ...options.command,
  ])
}

export const listWindowPanes = async (
  tmux: TmuxClient,
  paneId: string,
): Promise<WindowPaneContext[]> => {
  let output = ''

  try {
    output = (
      await tmux.capture([
        'list-panes',
        '-t',
        paneId,
        '-F',
        '#{pane_id}\t#{pane_in_mode}\t#{pane_width}\t#{pane_height}\t#{pane_current_path}\t#{pane_left}\t#{pane_top}\t#{?pane_active,1,0}\t#{window_zoomed_flag}',
      ])
    ).trim()
  } catch (error) {
    throw new Error('tmux-fuzzy-motion: pane not found', { cause: error })
  }

  const panes = output
    .split('\n')
    .filter((line) => line.length > 0)
    .map((line) => {
      const [
        resolvedPaneId,
        paneInMode,
        width,
        height,
        currentPath,
        left,
        top,
        active,
        zoomed,
      ] = line.split('\t')
      const numeric = [width, height, left, top].map((value) => Number(value))

      if (
        !resolvedPaneId ||
        !paneInMode ||
        !currentPath ||
        [active, zoomed].some((value) => value === undefined) ||
        numeric.some((value) => !Number.isFinite(value))
      ) {
        throw new Error('tmux-fuzzy-motion: failed to resolve window panes')
      }

      return {
        paneId: resolvedPaneId,
        inCopyMode: paneInMode === '1',
        width: Number(width),
        height: Number(height),
        currentPath,
        left: Number(left),
        top: Number(top),
        active: active === '1',
        zoomed: zoomed === '1',
      }
    })

  if (panes.length === 0) {
    throw new Error('tmux-fuzzy-motion: pane not found')
  }

  const zoomed = panes.some((pane) => pane.zoomed)
  return panes
    .filter((pane) => !zoomed || pane.active)
    .map((pane) => {
      const { zoomed: paneZoomed, ...rest } = pane
      void paneZoomed
      return rest
    })
}

export const getPaneBorderLines = async (
  tmux: TmuxClient,
  paneId: string,
): Promise<PaneBorderLines> => {
  let output = ''

  try {
    output = (
      await tmux.capture([
        'show-options',
        '-A',
        '-wv',
        '-t',
        paneId,
        'pane-border-lines',
      ])
    ).trim()
  } catch (error) {
    throw new Error('tmux-fuzzy-motion: failed to resolve pane border lines', {
      cause: error,
    })
  }

  if (
    output === 'single' ||
    output === 'double' ||
    output === 'heavy' ||
    output === 'simple' ||
    output === 'number' ||
    output === 'spaces'
  ) {
    return output
  }

  throw new Error('tmux-fuzzy-motion: failed to resolve pane border lines')
}

export const getTmuxVersion = async (): Promise<string> => {
  const result = await runProcess('tmux', ['-V'])
  return result.stdout.trim()
}

File: src/tmux.rs
use anyhow::{Context, Result, anyhow, bail};
use std::process::Command;

use crate::types::{PaneStartContext, WindowPaneContext};

pub fn run_process(command: &str, args: &[String]) -> Result<String> {
    let output = Command::new(command)
        .args(args)
        .output()
        .with_context(|| format!("failed to run {command}"))?;
    if !output.status.success() {
        bail!("{}", String::from_utf8_lossy(&output.stderr).trim());
    }
    Ok(String::from_utf8_lossy(&output.stdout).to_string())
}

pub fn tmux(args: &[String]) -> Result<String> {
    run_process("tmux", args)
}

pub fn tmux_quiet(args: &[String]) {
    let _ = tmux(args);
}

pub fn s(values: &[&str]) -> Vec<String> {
    values.iter().map(|value| (*value).to_string()).collect()
}

pub fn get_pane_start_context(pane_id: &str) -> Result<PaneStartContext> {
    let format = "#{pane_id}\t#{pane_in_mode}\t#{pane_width}\t#{pane_height}\t#{pane_current_path}";
    let output = tmux(&s(&["display-message", "-p", "-t", pane_id, format]))
        .map_err(|error| anyhow!("tmux-fuzzy-motion: pane not found: {error}"))?;
    let parts: Vec<&str> = output.trim().split('\t').collect();
    if parts.len() < 5 {
        bail!("tmux-fuzzy-motion: failed to resolve pane context");
    }
    Ok(PaneStartContext {
        pane_id: parts[0].to_string(),
        in_copy_mode: parts[1] == "1",
        width: parts[2].parse()?,
        height: parts[3].parse()?,
        current_path: parts[4].to_string(),
    })
}

pub fn list_window_panes(pane_id: &str) -> Result<Vec<WindowPaneContext>> {
    let format = "#{pane_id}\t#{pane_in_mode}\t#{pane_width}\t#{pane_height}\t#{pane_current_path}\t#{pane_left}\t#{pane_top}\t#{?pane_active,1,0}\t#{window_zoomed_flag}\t#{pane-border-lines}";
    let output = tmux(&s(&["list-panes", "-t", pane_id, "-F", format]))
        .map_err(|error| anyhow!("tmux-fuzzy-motion: pane not found: {error}"))?;
    let mut panes = Vec::new();
    let mut zoomed = false;
    for line in output.lines().filter(|line| !line.is_empty()) {
        let parts: Vec<&str> = line.split('\t').collect();
        if parts.len() < 10 {
            bail!("tmux-fuzzy-motion: failed to resolve window panes");
        }
        zoomed |= parts[8] == "1";
        panes.push((
            WindowPaneContext {
                pane_id: parts[0].to_string(),
                in_copy_mode: parts[1] == "1",
                width: parts[2].parse()?,
                height: parts[3].parse()?,
                current_path: parts[4].to_string(),
                left: parts[5].parse()?,
                top: parts[6].parse()?,
                active: parts[7] == "1",
                border_lines: parts[9].to_string(),
            },
            parts[8] == "1",
        ));
    }
    let filtered: Vec<WindowPaneContext> = panes
        .into_iter()
        .filter_map(|(pane, _)| (!zoomed || pane.active).then_some(pane))
        .collect();
    if filtered.is_empty() {
        bail!("tmux-fuzzy-motion: pane not found");
    }
    Ok(filtered)
}

pub fn enter_copy_mode(pane_id: &str) -> Result<()> {
    tmux(&s(&["copy-mode", "-t", pane_id]))
        .map_err(|error| anyhow!("tmux-fuzzy-motion: failed to enter copy-mode: {error}"))?;
    Ok(())
}

pub fn exit_copy_mode(pane_id: &str) -> Result<()> {
    tmux(&s(&["send-keys", "-X", "-t", pane_id, "cancel"]))
        .map_err(|error| anyhow!("tmux-fuzzy-motion: failed to exit copy-mode: {error}"))?;
    Ok(())
}
